plutus-core/testlib/PlutusIR/Properties/Typecheck.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+
+{- | Typechecking property tests for all PIR compiler passes. Currently we only typecheck the term
+ after a pass.
+
+ There were test failures when we infer the type before a pass and check that the
+ type after a pass is the same. The failures are probably caused by the
+ `PlutusIR.TypeCheck.AllowEscape` typechecker config.
+-}
+module PlutusIR.Properties.Typecheck where
+
+import Control.Monad.Reader
+import PlutusCore.Default
+import PlutusCore.Generators.QuickCheck.Utils
+import PlutusCore.Quote
+import PlutusCore.Rename
+import PlutusCore.TypeCheck qualified as PLC
+import PlutusIR.Compiler
+import PlutusIR.Compiler.Provenance (original)
+import PlutusIR.Core
+import PlutusIR.Generators.QuickCheck.GenerateTerms
+import PlutusIR.TypeCheck
+import PlutusPrelude (($>))
+import Test.QuickCheck (Property, elements)
+import Test.QuickCheck.Arbitrary
+
+instance Arbitrary (BuiltinSemanticsVariant DefaultFun) where
```

Comment:
you even wrote this but I don't think you're using it!

Reply:
It won't compile without it. And I'm not using it because it makes it hard to mark expect failures. Although I can use it for the ones without failures atm. But then I'm worried it's inconsistent/messy. What do you think?

Reply:
> It won't compile without it

That doesn't tell me anything - why doesn't it compile? I think you mean "we're using it somewhere", but where?

Not that it matters much, since I'm suggesting you _do_ use it. I think it's fine as an instance, I just think we should be exactly consistent and generate it everywhere.